### PR TITLE
Feature/add seeed grove gmxxx multichannel gas support

### PR DESCRIPTION
**Pull request by YorkshireIoT**
# What does this implement/fix?

Add support for Grove Multichannel Gas Sensor V2 (https://wiki.seeedstudio.com/Grove-Multichannel-Gas-Sensor-V2). The sensors themselves do not produce measurements with units like PPM but instead just raw values read from the sensor ADC. Ideally there would be some baselining built into the sensor but I don't have any other sensors I could provide and test that baselining against. Hopefully this is something someone else can add or I can add further down the line. The raw values still provide insight into relative local changes in the different gas concentrations. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2598

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: grove_gas_mc_v2
    no2:
      name: "Nitrogen Dioxide"
    ethanol:
      name: "Ethanol"
    carbon_monoxide:
      name: "Carbon Monoxide"
    tvoc:
      name: "Volatile Organic Compounds"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/grove_gas_mc_v2/grove_gas_mc_v2.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+static const uint8_t GROVE_GAS_MC_V2_READ_GM302B = 0x03;
+static const uint8_t GROVE_GAS_MC_V2_READ_GM502B = 0x05;
+static const uint8_t GROVE_GAS_MC_V2_READ_GM702B = 0x07;
+
+#define CHECK_TRUE(f, error_code) \
+  if (!(f)) { \
+    this->mark_failed(); \
+    this->error_code_ = (error_code); \
+    return; \
+  }
```
Comment (jesserockz) [inline esphome/components/grove_gas_mc_v2/grove_gas_mc_v2.cpp]: No need to create a macro that is only ever used once

File: esphome/components/grove_gas_mc_v2/sensor.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    ICON_GAS_CYLINDER,
+    UNIT_EMPTY,
+)
+
+DEPENDENCIES = ["i2c"]
+CODEOWNERS = ["@YorkshireIoT"]
+
+# Default component values
+DEFAULT_I2C_ADDRESS = 0x08
+DEFAULT_POLLING_FREQUENCY = "60s"
```
Comment (jesserockz) [inline esphome/components/grove_gas_mc_v2/sensor.py]: Please just put these inline to match the rest of the codebase

**Review by jesserockz — CHANGES_REQUESTED** [review]
Please remove unrelated formatting changes in the test yaml file.
The preferred is actually to have no quote around strings that do not have special characters.

**Comment by YorkshireIoT** [comment]
> Please remove unrelated formatting changes in the test yaml file.

@jesserockz 
R.e. the formatting changes in `test1.yaml`, these were done automatically on save within the dev container by the YAML extension and it actually stops you  from reverting the changes while the extension is still enabled:
https://user-images.githubusercontent.com/55233103/221353135-76a5d21b-706c-4c01-a564-d72dbfeb63f2.mp4

I had to disable the extension to be able to save the file and commit/push the changes

**Comment by YorkshireIoT** [comment]
`lint-python` failing due to unrelated error in `esphome/components/esp32/__init__.py`, but this is fixed in [PR#4479](https://github.com/esphome/esphome/pull/4479/files)

**Comment by jesserockz** [comment]
> I had to disable the extension to be able to save the file and commit/push the changes

You can save without auto formatting with `ctrl+k, s`

**Comment by YorkshireIoT** [comment]
> > I had to disable the extension to be able to save the file and commit/push the changes
> 
> You can save without auto formatting with `ctrl+k, s`

Ah thanks, that's good to know

**Comment by YorkshireIoT** [comment]
Hi @jesserockz, could I get a rereview of the changes you requested on this please?

**Comment by YorkshireIoT** [comment]
Hi @jesserockz,

I've tidied up the commits and rebased and resolved some merge conflicts with unrelated changes that have gone into dev since I first open this PR.

This is ready for a rereview when you have time.

Thanks 😄

**Comment by YorkshireIoT** [comment]
Workflow run failing due to a 429 Git checkout error. If someone with the right priviledges could re-run the job when they get a chance that'd be really appreciated 😄

**Comment by YorkshireIoT** [comment]
Ping

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**